csrc/kernel_cache.cpp

```diff
@@ -201,7 +222,55 @@ std::vector<at::Tensor> FusionExecutorCache::runFusionWithInputs(
 
   KernelArgumentHolder args = prepareInputs(perm_inputs);
 
-  auto kernel_runtime = getKernelRuntimeFor(args, forced_index_type);
+  // Concretizing the fusion using args
+  auto fusion = fusion_.get();
+  if (has_dynamic_reshape_) {
+    ExpressionEvaluator expr_eval;
+
+    // Bind input scalars and tensor metadata to symbolic scalars
+    // Here we bind only the inputs that are needed to concretize dynamic
+    // transforms.
+    TORCH_CHECK(
+        args.size() == fusion->inputs().size(),
+        "Received ",
+        args.size(),
+        " inputs but expected ",
+        fusion->inputs().size());
+    for (auto i : c10::irange(args.size())) {
+      auto inpi = fusion->inputs()[i];
+      if (inpi->isIntegralScalar() &&
+          concretization_inputs_.find(inpi) != concretization_inputs_.end()) {
+        // We should already have checked that inputs[i] is the appropriate
+        // type when running prepareInputs.
+        expr_eval.bind(inpi, inputs[i].toInt());
+      } else if (inpi->isA<TensorView>()) {
+        auto tv = inpi->as<TensorView>();
+        auto root = tv->domain()->getRootDomain();
+        for (auto j : c10::irange(root.size())) {
+          auto id = root[j];
+          if (concretization_inputs_.find(id->extent()) !=
+              concretization_inputs_.end()) {
+            expr_eval.bind(id->extent(), (int64_t)inputs[i].toTensor().size(j));
+          }
+        }
+      }
+    }
+
+    auto conc_inf = DynamicTransform::getConcretizationInfo(fusion, &expr_eval);
+    auto conc_it = concretized_fusions_.find(conc_inf);
+    if (conc_it != concretized_fusions_.end()) {
+      fusion = conc_it->second.get();
+    } else { // cache miss
+      auto res = concretized_fusions_
+                     .emplace(conc_inf, std::make_unique<Fusion>(*fusion))
+                     .first;
+      fusion = res->second.get();
+      DynamicTransform::concretizeFusion(fusion, conc_inf);
+    }
+  }
+  FusionGuard fg(fusion);
+
+  auto kernel_runtime = getKernelRuntimeFor(args, forced_index_type, fusion);
```

Comment:
nitpick: leaning towards having the check for concretizationInfo inside `getkernelRuntimeFor` as well. Since the whole concretization info is also specialized by kernel runtime.